activerecord/lib/active_record/attribute_methods/primary_key.rb

```diff
@@ -16,34 +16,32 @@ def to_key
 
       # Returns the primary key column's value.
       def id
-        primary_key = self.class.primary_key
-        _read_attribute(primary_key) if primary_key
+        _read_attribute(@primary_key)
       end
 
       # Sets the primary key column's value.
       def id=(value)
-        primary_key = self.class.primary_key
-        _write_attribute(primary_key, value) if primary_key
+        _write_attribute(@primary_key, value)
```

Comment:
Previously `#id=` behaves as no-op if primary key doesn't exist.
But I think that calling `#id=` on no-pk object is already something wrong, raising `MissingAttributeError` is preferable than ignore any error to me.

And also, since most models has pk, this guard is almost redundant.
E.g. `persistence.rb#L932` checks `if @primary_key` three times as before, now only once.

https://github.com/rails/rails/blob/b6828fc91531ae0cc0a0f216705dd19112596301/activerecord/lib/active_record/persistence.rb#L928-L932